src/workerd/api/web-socket.c++

```diff
@@ -344,23 +353,31 @@ kj::Promise<DeferredProxy<void>> WebSocket::couple(kj::Own<kj::WebSocket> other)
                                           downstream.eagerlyEvaluate(nullptr)))
       .attach(kj::mv(self), kj::mv(other));
 
-  if (locality == LOCAL) {
-    // Since the WebSocket is terminated locally, we need the IoContext to stay live while
-    // it is active.
-    co_await promise;
-    co_return;
-  } else {
-    // Since the WebSocket is just proxying through, we can do the pump in a deferred proxy task.
-    // Note that we don't need to (and can't) register any pending events in this case since the
-    // IoContext is free to go away at this point.
+  KJ_IF_SOME(peer, tryGetPeer()) {
+    // Since the WebSocket is terminated locally, we generally want the request and associated
+    // IoContext to stay alive until the WebSocket connection has terminated.
+    //
+    // However, there is one exception to this: when the WebSocket is hibernatable, we don't want
+    // the existence of this connection to prevent the actor from being evicted, so we fall through
+    // to deferred proxying in this case.
+    if (!isHibernatable(peer)) {
+      co_await promise;
+      co_return;
+    }
+  }
 
-    // To begin deferred proxying, we can use this magic `KJ_CO_MAGIC` expression, which fulfills
-    // our outer promise for a DeferredProxy<void>, which wraps a promise for the rest of this
-    // coroutine.
-    KJ_CO_MAGIC BEGIN_DEFERRED_PROXYING;
+  // Either:
+  // 1. This websocket is just proxying through, in which case we can allow the IoContext to go
+  // away while still being able to successfully pump the websocket connection.
+  // 2. This is a hibernatable websocket and we are falling through to deferred proxying to
```

Comment:
FWIW this approach of treating hibernatable websockets as "deferred-proxyable" feels a little bit hacky and I'm inclined to suggest that the problem it solves should be solved outside this class. However, this is intended to be an NFC so I'm not looking to change this behaviour here.